Added support for naming mocked functions

gricard opened the pull request:
**Summary**

This patch adds the ability to specify a name for a just.fn() mock, so if the test fails, you will see that name in the error output instead of "just.fn()", which should help make some test results clearer. It provides a few different ways to specify the mock name.

**Motivation** 

I switched a frontend product over to Jest from Karma/Mocha/Chai a few months ago, and I love it. But there's a couple things that I wanted to see in the output, like specified mock names, and I did not see any existing means to add that to my tests. 

**Test plan**

Unit tests and integration tests have been added.

SimenB commented:
@gricard You can add an integration test, there you can get both stdout and stderr. Lots of examples in there

gricard commented:
Thanks, @SimenB. Am I able to modify this PR once I figure out how to add those tests? Or do I need to do a new one?

gricard commented:
Interesting... AppVeyor build failed, and it looks like it's because the snapshot tests failed due to slight differences in the characters used. Should I not have committed the snapshots created on my Mac? That appears to be the problem. Output differences between macOS / Windows. 

```
    Received value does not match stored snapshot 1.
    
    - Snapshot
    + Received
    
    - "PASS __tests__/index.js
    -   ✓ first test
    + "PASS __tests__\\index.js
    +   √ first test
      
      "
```

gricard commented:
Well, if the test output varies by platform, then I'll do regex matching instead

SimenB commented:
You can do `skipOnWindows`, not sure if that's wanted here, though

SimenB commented on docs/en/MockFunctionAPI.md:
````diff
@@ -138,13 +138,32 @@ console.log(myMockFn(), myMockFn(), myMockFn(), myMockFn());
 > 'first call', 'second call', 'default', 'default'
 ```
 
-### `mockFn.mockReturnThis()`
````
why is this removed?

gricard replied:
Looks like an accident on my part. I'll fix it.

SimenB commented on integration_tests/mock-names/with-empty-mock-name/__tests__/index.js:
```diff
@@ -0,0 +1,19 @@
+test('first test', () => {
+  mockFn();
+  expect(mockFn).toHaveBeenCalled();
+  expect(mockFn.mock.calls.length).toBe(1);
```
any particular reason why you do this instead of `expect(mockFn)`.toHaveBeenCalledTimes(1)`?

gricard replied:
No reason. I just copied what was in another test suite in there that I based mine on.

SimenB commented on integration_tests/__tests__/mock_names.test.js:
```diff
@@ -0,0 +1,70 @@
+/**
```
Can you add a test for failing `expect(mockWithName).not.toHaveBeenCalled()` and `xpect(mockWithName).toHaveBeenCalledTimes(5)` to assert on the name for them as well? I see no reason for them not to work as you expect, but having the assertions would be nice

gricard replied:
So, a test where the assertion `expect(mockWithName).not.toHaveBeenCalled()` will fail, and then check for that error?

And a test for `expect(mockWithName).toHaveBeenCalledTimes(5)` when it's not called 5 times?

gricard replied:
Ok, I've added more tests to cover those.

SimenB replied:
Yup. Thanks!

cpojer commented:
I quite like it. Thanks for contributing this change to Jets. However, I'd prefer to not add an argument to `jest.fn` at this point. Could we change this PR to only add `.mockName`? Thank you.

gricard commented:
Whatever floats your boat.

gricard commented:
Hm. Well, I made the changes, but circleci test-node-8 is failing because of an unrelated test failure. 

```
 FAIL  integration_tests/__tests__/transform.test.js (24.846s)
  ● babel-jest › runs transpiled code

    
          ORIGINAL CMD: yarn
          STDOUT: 
          STDERR: warning package.json: No license field
    warning ../../package.json: No license field
    warning No license field
    error An unexpected error occurred: "https://registry.yarnpkg.com/core-js/-/core-js-2.4.1.tgz: Request failed \"530 undefined\"".
    
          STATUS: 1
          ERROR: undefined
      
          
      at run (integration_tests/utils.js:33:11)
      at Object.beforeEach (integration_tests/__tests__/transform.test.js:28:7)
          at Promise (<anonymous>)
          at <anonymous>
```

SimenB commented:
Restarted it, and CI is happy now

gricard commented:
Sweet. Thanks, dude.

SimenB commented:
Question: Is the name cleared by calling `mock.mockReset()`? Should it be?

gricard commented:
Ooh. Yeah. That makes sense.  I didn't think of that. I can change it so that clears out the name also.

SimenB commented:
I think it should `mockReset` and `mockRestore` should clear it, but not `mockClear`.

gricard commented:
I took a look at the docs for all of those. I think only mockReset() makes sense, in terms of clearing out the name. mockRestore() is specific to the mocked implementation.

cpojer commented:
Yep, agreed. :)

gricard commented:
It looks like mockReset() already does this, because it clears out the mock entirely with its config (where the mock name is stored). I've added a unit test to verify this, and updated the documentation for mockReset().

SimenB commented:
For funsies, can you add a test that verifies that `mock{Clear,Restore}` does _not_ clear the name?

gricard commented:
All set, @SimenB

gricard commented:
And thanks! I really like Jest and this will help me out a lot as well. Hopefully I can contribute some more to the project. :)

cpojer commented:
Alright, I think this is almost good to go, I would just like to request one more change (sorry :( ). Could we split the function up into `mockName` as a setter and `getMockName` to get the value? I generally don't like functions that behave differently based on the amount of arguments. Thank you.

gricard commented:
All set

cpojer commented:
Thanks for sticking with us, this is a great PR :)

gricard commented:
Thanks, dude!

SimenB commented on docs/en/MockFunctions.md:
```diff
@@ -217,6 +217,19 @@ const otherObj = {
+  .mockImplementation(scalar => 42 + scalar)
+  .mockName('add42');
+
+const myMockFn2 = jest.fn(scalar => 42 + scalar, 'add42');
```
this is wrong now

gricard replied:
Damn. I thought I caught all the doc references. Can I still commit changes for this PR, or is that done since it's merged?

SimenB replied:
New PR is needed

SimenB commented on docs/en/MockFunctions.md:
````diff
@@ -217,6 +217,19 @@ const otherObj = {
 };
 ```
 
+## Mock Names
````
should it include a "availble in jest 21.4" or something?
